drivers/docker/driver.go

```diff
@@ -0,0 +1,1682 @@
+	}
+
+	return nil
+}
+
+func (d *Driver) WaitTask(ctx context.Context, taskID string) (<-chan *drivers.ExitResult, error) {
+	h, ok := d.tasks.Get(taskID)
+	if !ok {
+		return nil, drivers.ErrTaskNotFound
+	}
+	ch := make(chan *drivers.ExitResult)
+	go d.handleWait(ctx, ch, h)
+	return ch, nil
+}
+
+func (d *Driver) handleWait(ctx context.Context, ch chan *drivers.ExitResult, h *taskHandle) {
+	defer close(ch)
+	select {
+	case <-h.waitCh:
+		ch <- h.exitResult
+	case <-ctx.Done():
+		ch <- &drivers.ExitResult{
+			Err: ctx.Err(),
+		}
+	}
+}
+
+func (d *Driver) StopTask(taskID string, timeout time.Duration, signal string) error {
+	h, ok := d.tasks.Get(taskID)
+	if !ok {
+		return drivers.ErrTaskNotFound
+	}
+
+	if signal == "" {
+		signal = "SIGINT"
+	}
+
+	sig, err := signals.Parse(signal)
+	if err != nil {
+		return fmt.Errorf("failed to parse signal: %v", err)
+	}
+
+	return h.Kill(timeout, sig)
+}
+
+func (d *Driver) DestroyTask(taskID string, force bool) error {
+	h, ok := d.tasks.Get(taskID)
+	if !ok {
+		return drivers.ErrTaskNotFound
+	}
+
+	defer h.dloggerPluginClient.Kill()
+
+	c, err := h.client.InspectContainer(h.container.ID)
+	if err != nil {
+		return fmt.Errorf("failed to inspect container state: %v", err)
+	}
+	if c.State.Running && !force {
+		return fmt.Errorf("must call StopTask for the given task before Destroy or set force to true")
+	}
+
+	if err := h.client.StopContainer(h.container.ID, 0); err != nil {
+		h.logger.Warn("failed to stop container during destroy", "error", err)
+	}
+
+	if err := h.dlogger.Stop(); err != nil {
+		h.logger.Error("failed to stop docker logger process during destroy",
+			"error", err, "logger_pid", h.dloggerPluginClient.ReattachConfig().Pid)
+	}
+
+	if err := d.cleanupImage(h); err != nil {
+		h.logger.Error("failed to cleanup image after destroying container",
+			"error", err)
+	}
+
+	return nil
+}
+
+// cleanupImage removes a Docker image. No error is returned if the image
+// doesn't exist or is still in use. Requires the global client to already be
+// initialized.
+func (d *Driver) cleanupImage(handle *taskHandle) error {
+	if !d.config.ImageGC {
+		return nil
+	}
+
+	coordinator, callerID := d.getDockerCoordinator(client, handle.task)
+	coordinator.RemoveImage(handle.container.Image, callerID)
+
+	return nil
+}
+
+func (d *Driver) InspectTask(taskID string) (*drivers.TaskStatus, error) {
+	h, ok := d.tasks.Get(taskID)
+	if !ok {
+		return nil, drivers.ErrTaskNotFound
+	}
+
+	status := &drivers.TaskStatus{
+		ID:          h.task.ID,
+		Name:        h.task.Name,
+		StartedAt:   h.container.State.StartedAt,
+		CompletedAt: h.container.State.FinishedAt,
+		DriverAttributes: map[string]string{
+			"container_id": h.container.ID,
+		},
+		NetworkOverride: h.net,
+		ExitResult:      h.exitResult,
+	}
+
+	status.State = drivers.TaskStateUnknown
+	if h.container.State.Running {
+		status.State = drivers.TaskStateRunning
+	}
+	if h.container.State.Dead {
```

Comment:
These checks seem racy. Have you run tests with -race?

Reply:
Yup, no race detected! Doesn't mean they aren't racy. Under what conditions do you see a race?

Reply:
Can't a task exit concurrently with InspectTask and write to `h.container.State` while this is reading it?

Reply:
Ah I see what you mean. This shouldn't be referencing `h.container` at all actually and should instead just call `docker.ContainerInspect`. Will fix.

Reply:
Take a look at https://github.com/hashicorp/nomad/pull/4844/commits/105025b4499f492bdc27cd3226bef2bae144ddf8

`h.container` was only written once when the handle was created. Only the container ID and Image were ever referenced so I pulled those out into their own fields on the handle.